### PR TITLE
Unify hasLabel() and labelContains()

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

**Breaking change:** Policy rules which use `labelContains()` must be changed to use `hasLabel()` instead.

### CODE REVIEW

**Review by MarcelBochtler — DISMISSED** [review]

File: evaluator/src/main/kotlin/Rule.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/Rule.kt:109]: Maybe simply: ".. a label named [key] ..."

File: evaluator/src/main/kotlin/Rule.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/Rule.kt:110]: I believe that description is misleading now. It sounds like if the label's value is e.g. a comma-separated string, then that string is interpreted as a list, and it is checked whether `value` is part of that list. But that's not how it's implemented.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: evaluator/src/main/kotlin/Rule.kt
Comment (sschuberth) [reply evaluator/src/main/kotlin/Rule.kt:110]: > I wonder whether there is actually any use case for `hasLabel()`.

I believe there is. Lazy people might simply want to check for the presence of a `DO_SOMETHING` label instead of requiring the value to be "true" / "false" and parsing it as a boolean.

Comment (fviernau) [reply evaluator/src/main/kotlin/Rule.kt:110]: Let me try to clarify what I meant. 

Can we replace labelContains() and hasLabel() (see Rule.kt) with the following one in OrtResult plus 
an identical one (which returns a matcher instead of bool) in the Rule class?
```
    fun hasLabel(key: String, value: String? = null) =
        if (value == null) {
            key in labels
        } else {
            value in getLabelValues(key)
        }
```

Comment (fviernau) [reply evaluator/src/main/kotlin/Rule.kt:110]: @sschuberth I've extracted the IIUC uncontroversial bit to https://github.com/oss-review-toolkit/ort/pull/6084 to unblock me.
Once this is merged, I'll make a force push to this PR proposing something on top.

Comment (mnonnenmacher) [reply evaluator/src/main/kotlin/Rule.kt:110]: > Lazy people might simply want to check for the presence of a DO_SOMETHING label instead of requiring the value to be "true" / "false" and parsing it as a boolean.

It has nothing to do with being lazy, but yes, a label does not require to have a value.

> Can we replace labelContains() and hasLabel() (see Rule.kt) with the following one...

That would make it impossible to check the exact value of a label, because it is always interpreted as a comma-separated list. Currently the user can decide how the value is interpreted so I think I would prefer to keep it as is.

File: evaluator/src/main/kotlin/Rule.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/Rule.kt:112]: I prefer the old name because the string before the `=` is the actual label.

File: evaluator/src/main/kotlin/Rule.kt
Comment (sschuberth) [reply evaluator/src/main/kotlin/Rule.kt:110]: > It has nothing to do with being lazy, but yes, a label does not require to have a value.

I should have put "lazy" in quotes as it was a flippant remark to be taken with a wink.

> Currently the user can decide how the value is interpreted so I think I would prefer to keep it as is.

I tend to agree, although it's probably quite unlike that a user is using commas in the value and these commas are *not* list item delimiters.

Comment (fviernau) [reply evaluator/src/main/kotlin/Rule.kt]: I've updated it and hope it now fits the preferences of both of you @mnonnenmacher @sschuberth .

File: evaluator/src/main/kotlin/Rule.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/Rule.kt]: Looking at this commit only, it's unclear what the benefit is to use `hasLabel` instead of `getLabelValues` in `labelContains`. And it's probably hard to find an argument for the commit message, as the next commit even removes `labelContains`. So why bother changing it here in the first place?

Comment (sschuberth) [reply evaluator/src/main/kotlin/Rule.kt]: Boolean arguments should be named for clarity.

File: evaluator/src/test/kotlin/RuleTest.kt
Comment (sschuberth) [inline evaluator/src/test/kotlin/RuleTest.kt:167]: The boolean argument should be named.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]

**Pull request merged**
